src/builder-source-archive.c

```diff
@@ -579,11 +591,38 @@ builder_source_archive_extract (BuilderSource  *source,
   g_autofree char *archive_path = NULL;
   BuilderArchiveType type;
   gboolean is_local;
+  g_autoptr(GFile) dest_file = NULL;
+  g_autofree char *dest_filename = NULL;
 
   archivefile = get_source_file (self, context, &is_local, error);
   if (archivefile == NULL)
     return FALSE;
 
+  if (self->dest_filename)
+    {
+      dest_filename = g_strdup (self->dest_filename);
+      dest_file = g_file_get_child (dest, dest_filename);
+
+      /* If the destination file exists, just delete it. We can encounter errors when
+       * trying to overwrite files that are not writable. */
+      if (g_file_query_exists (dest_file, NULL) && !g_file_delete (dest_file, NULL, error))
+        return FALSE;
+
+      /* Make sure the target is gone, because g_file_copy does
+         truncation on hardlinked destinations */
+      (void)g_file_delete (dest_file, NULL, NULL);
+
+      if (!g_file_copy (archivefile, dest_file,
+            G_FILE_COPY_OVERWRITE,
+            NULL,
+            NULL, NULL,
+            error))
+        return FALSE;
```

Comment:
This downloads the file to the original name and then makes a copy. It seems to me that it would make more sense to just download it to the right name the first time. I.e. in get_download_location() use dest_filename for  "basename"

Reply:
OK, will fix it.